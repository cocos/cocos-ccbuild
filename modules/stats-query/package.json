{
  "name": "@ccbuild/stats-query",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "build": "tsc && npx ts-json-schema-generator -t Config -p ./src/config-interface.ts -o ./lib/cc.config.schema.json"
  },
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "./lib/**/*"
  ],
  "exports": {
    ".": {
      "types": "./src/index.ts",
      "default": "./lib/index.js"
    }
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@ccbuild/modularize": "workspace:*",
    "@types/resolve": "^1.20.2",
    "dedent": "^0.7.0",
    "fs-extra": "~11.1.1",
    "glob": "~7.2.0",
    "json5": "^2.2.3",
    "resolve": "^1.22.1"
  },
  "devDependencies": {
    "@types/dedent": "^0.7.0",
    "@types/fs-extra": "^9.0.13",
    "@types/glob": "~7.2.0",
    "@types/jest": "^29.2.3",
    "@types/node": "~20.3.3",
    "del": "^6.1.1",
    "jest": "^29.3.1",
    "ts-jest": "^29.0.3",
    "ts-json-schema-generator": "^1.3.0"
  }
}
