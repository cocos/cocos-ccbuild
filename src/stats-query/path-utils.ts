import * as ps from 'path';

export function normalizePath (path: string): string {
    return path.replace(/\\/g, '/');
}

export function toExtensionLess (path: string) {
    return path.slice(0, -ps.extname(path).length);
}