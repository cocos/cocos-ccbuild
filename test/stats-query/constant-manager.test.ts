import { StatsQuery } from '@ccbuild/stats-query';
import ps from 'path';

const ConstantManager = StatsQuery.ConstantManager;
const cm = new ConstantManager(ps.join(__dirname, '../test-engine-source/'));

test('generateInternalConstants', () => {
  expect(cm.genInternalConstants()).toMatchSnapshot();
});

test('generateCCEnv', () => {
  expect(cm.genCCEnv()).toMatchSnapshot();
});

test('genCCEnvConstants', () => {
    expect(cm.genCCEnvConstants({
        mode: 'TEST',
        platform: 'NATIVE',
        flags: { DEBUG: false, SERVER_MODE: true },
      })).toMatchSnapshot();
});

test('exportStaticConstants', () => {
  expect(
    cm.exportStaticConstants({
      mode: 'PREVIEW',
      platform: 'WECHAT',
      flags: { DEBUG: false, SERVER_MODE: true },
    })
  ).toMatchSnapshot();
});

test('exportDynamicConstants', () => {
  expect(
    cm.exportDynamicConstants({
      mode: 'BUILD',
      platform: 'WECHAT',
      flags: { DEBUG: true, SERVER_MODE: true },
    })
  ).toMatchSnapshot();
});

test('genBuildTimeConstants', () => {
  const result = cm.genBuildTimeConstants({
    mode: 'TEST',
    platform: 'NATIVE',
    flags: { DEBUG: false, SERVER_MODE: true },
  });
  expect(result).toMatchSnapshot();
});
