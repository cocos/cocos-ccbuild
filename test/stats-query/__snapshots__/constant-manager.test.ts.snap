// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exportDynamicConstants 1`] = `
"function tryDefineGlobal (name, value) {
    const _global = typeof window === 'undefined' ? global : window;
    if (typeof _global[name] === 'undefined') {
        return (_global[name] = value);
    } else {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return _global[name];
    }
}
function defined (name) {
    const _global = typeof window === 'undefined' ? global : window;
    return typeof _global[name] === 'object';
}

export const TEST = tryDefineGlobal('CC_TEST', defined('tap') || defined('QUnit'));
export const EDITOR = tryDefineGlobal('CC_EDITOR', defined('Editor') && defined('process') && ('electron' in process.versions));
export const PREVIEW = tryDefineGlobal('CC_PREVIEW', !EDITOR);
export const JSB = tryDefineGlobal('CC_JSB', defined('jsb'));
export const NATIVE = JSB;
export const HTML5 = !(EDITOR && NATIVE);
export const DEV = tryDefineGlobal('CC_DEV', true);
export const EDITOR_NOT_IN_PREVIEW = EDITOR && !tryDefineGlobal('isPreviewProcess',false);

export const ANDROID = false;

export const IOS = false;

export const MAC = false;

export const WINDOWS = false;

export const LINUX = false;

export const OHOS = false;

export const OPEN_HARMONY = false;

export const WECHAT = true;
tryDefineGlobal('CC_WECHAT', true);

export const WECHAT_MINI_PROGRAM = false;

export const BAIDU = false;
tryDefineGlobal('CC_BAIDU', false);

export const XIAOMI = false;
tryDefineGlobal('CC_XIAOMI', false);

export const ALIPAY = false;
tryDefineGlobal('CC_ALIPAY', false);

export const TAOBAO = false;

export const TAOBAO_MINIGAME = false;

export const BYTEDANCE = false;
tryDefineGlobal('CC_BYTEDANCE', false);

export const OPPO = false;
tryDefineGlobal('CC_OPPO', false);

export const VIVO = false;
tryDefineGlobal('CC_VIVO', false);

export const HUAWEI = false;
tryDefineGlobal('CC_HUAWEI', false);

export const COCOSPLAY = false;
tryDefineGlobal('CC_COCOSPLAY', false);

export const QTT = false;
tryDefineGlobal('CC_QTT', false);

export const LINKSURE = false;
tryDefineGlobal('CC_LINKSURE', false);

export const BUILD = true;
tryDefineGlobal('CC_BUILD', true);

export const DEBUG = true;
tryDefineGlobal('CC_DEBUG', true);

export const SERVER_MODE = true;

export const MINIGAME = true;
tryDefineGlobal('CC_MINIGAME', true);

export const RUNTIME_BASED = false;
tryDefineGlobal('CC_RUNTIME_BASED', false);

export const SUPPORT_JIT = false;
tryDefineGlobal('CC_SUPPORT_JIT', false);

export const NOT_PACK_PHYSX_LIBS = false;

export const NET_MODE = 0;

export const WEBGPU = false;

export const NATIVE_CODE_BUNDLE_MODE = 2;

export const WASM_SUBPACKAGE = false;

export const CULL_MESHOPT = false;

"
`;

exports[`exportStaticConstants 1`] = `
"function tryDefineGlobal (name, value) {
    const _global = typeof window === 'undefined' ? global : window;
    if (typeof _global[name] === 'undefined') {
        return (_global[name] = value);
    } else {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return _global[name];
    }
}
export const HTML5 = false;

export const NATIVE = false;

export const ANDROID = false;

export const IOS = false;

export const MAC = false;

export const WINDOWS = false;

export const LINUX = false;

export const OHOS = false;

export const OPEN_HARMONY = false;

export const WECHAT = true;
tryDefineGlobal('CC_WECHAT', true);

export const WECHAT_MINI_PROGRAM = false;

export const BAIDU = false;
tryDefineGlobal('CC_BAIDU', false);

export const XIAOMI = false;
tryDefineGlobal('CC_XIAOMI', false);

export const ALIPAY = false;
tryDefineGlobal('CC_ALIPAY', false);

export const TAOBAO = false;

export const TAOBAO_MINIGAME = false;

export const BYTEDANCE = false;
tryDefineGlobal('CC_BYTEDANCE', false);

export const OPPO = false;
tryDefineGlobal('CC_OPPO', false);

export const VIVO = false;
tryDefineGlobal('CC_VIVO', false);

export const HUAWEI = false;
tryDefineGlobal('CC_HUAWEI', false);

export const COCOSPLAY = false;
tryDefineGlobal('CC_COCOSPLAY', false);

export const QTT = false;
tryDefineGlobal('CC_QTT', false);

export const LINKSURE = false;
tryDefineGlobal('CC_LINKSURE', false);

export const EDITOR = false;
tryDefineGlobal('CC_EDITOR', false);

export const PREVIEW = true;
tryDefineGlobal('CC_PREVIEW', true);

export const BUILD = false;
tryDefineGlobal('CC_BUILD', false);

export const TEST = false;
tryDefineGlobal('CC_TEST', false);

export const DEBUG = false;
tryDefineGlobal('CC_DEBUG', false);

export const SERVER_MODE = true;

export const DEV = true;
tryDefineGlobal('CC_DEV', true);

export const MINIGAME = true;
tryDefineGlobal('CC_MINIGAME', true);

export const RUNTIME_BASED = false;
tryDefineGlobal('CC_RUNTIME_BASED', false);

export const SUPPORT_JIT = false;
tryDefineGlobal('CC_SUPPORT_JIT', false);

export const JSB = false;
tryDefineGlobal('CC_JSB', false);

export const NOT_PACK_PHYSX_LIBS = false;

export const NET_MODE = 0;

export const WEBGPU = false;

export const NATIVE_CODE_BUNDLE_MODE = 2;

export const EDITOR_NOT_IN_PREVIEW = false;

export const WASM_SUBPACKAGE = false;

export const CULL_MESHOPT = false;

"
`;

exports[`genBuildTimeConstants 1`] = `
{
  "ALIPAY": false,
  "ANDROID": false,
  "BAIDU": false,
  "BUILD": false,
  "BYTEDANCE": false,
  "COCOSPLAY": false,
  "CULL_MESHOPT": false,
  "DEBUG": false,
  "DEV": true,
  "EDITOR": false,
  "EDITOR_NOT_IN_PREVIEW": false,
  "HTML5": false,
  "HUAWEI": false,
  "IOS": false,
  "JSB": true,
  "LINKSURE": false,
  "LINUX": false,
  "MAC": false,
  "MINIGAME": false,
  "NATIVE": true,
  "NATIVE_CODE_BUNDLE_MODE": 2,
  "NET_MODE": 0,
  "NOT_PACK_PHYSX_LIBS": false,
  "OHOS": false,
  "OPEN_HARMONY": false,
  "OPPO": false,
  "PREVIEW": false,
  "QTT": false,
  "RUNTIME_BASED": false,
  "SERVER_MODE": true,
  "SUPPORT_JIT": true,
  "TAOBAO": false,
  "TAOBAO_MINIGAME": false,
  "TEST": true,
  "VIVO": false,
  "WASM_SUBPACKAGE": false,
  "WEBGPU": false,
  "WECHAT": false,
  "WECHAT_MINI_PROGRAM": false,
  "WINDOWS": false,
  "XIAOMI": false,
}
`;

exports[`genCCEnvConstants 1`] = `
{
  "ALIPAY": false,
  "ANDROID": false,
  "BAIDU": false,
  "BUILD": false,
  "BYTEDANCE": false,
  "COCOSPLAY": false,
  "DEBUG": false,
  "DEV": true,
  "EDITOR": false,
  "EDITOR_NOT_IN_PREVIEW": false,
  "HTML5": false,
  "HUAWEI": false,
  "IOS": false,
  "JSB": true,
  "LINKSURE": false,
  "LINUX": false,
  "MAC": false,
  "MINIGAME": false,
  "NATIVE": true,
  "NET_MODE": 0,
  "OHOS": false,
  "OPEN_HARMONY": false,
  "OPPO": false,
  "PREVIEW": false,
  "QTT": false,
  "RUNTIME_BASED": false,
  "SUPPORT_JIT": true,
  "TAOBAO": false,
  "TAOBAO_MINIGAME": false,
  "TEST": true,
  "VIVO": false,
  "WECHAT": false,
  "WECHAT_MINI_PROGRAM": false,
  "WINDOWS": false,
  "XIAOMI": false,
}
`;

exports[`generateCCEnv 1`] = `
"declare module 'cc/env'{
	/**
	 * Running in Web platform
	 */
	export const HTML5: boolean;

	/**
	 * Running in native platform (mobile app, desktop app, or simulator).
	 */
	export const NATIVE: boolean;

	/**
	 * Running in ANDROID platform
	 */
	export const ANDROID: boolean;

	/**
	 * Running in IOS platform
	 */
	export const IOS: boolean;

	/**
	 * Running in MAC platform
	 */
	export const MAC: boolean;

	/**
	 * Running in WINDOWS platform
	 */
	export const WINDOWS: boolean;

	/**
	 * Running in LINUX platform
	 */
	export const LINUX: boolean;

	/**
	 * Running in OHOS platform
	 */
	export const OHOS: boolean;

	/**
	 * Running in OPEN_HARMONY platform
	 */
	export const OPEN_HARMONY: boolean;

	/**
	 * Running in the Wechat's mini game.
	 */
	export const WECHAT: boolean;

	/**
	 * Running in the Wechat's mini program.
	 */
	export const WECHAT_MINI_PROGRAM: boolean;

	/**
	 * Running in the baidu's mini game.
	 */
	export const BAIDU: boolean;

	/**
	 * Running in the xiaomi's quick game.
	 */
	export const XIAOMI: boolean;

	/**
	 * Running in the alipay's mini game.
	 */
	export const ALIPAY: boolean;

	/**
	 * Running in the taobao creative app.
	 */
	export const TAOBAO: boolean;

	/**
	 * Running in the taobao mini game.
	 */
	export const TAOBAO_MINIGAME: boolean;

	/**
	 * Running in the ByteDance's mini game.
	 */
	export const BYTEDANCE: boolean;

	/**
	 * Running in the oppo's quick game.
	 */
	export const OPPO: boolean;

	/**
	 * Running in the vivo's quick game.
	 */
	export const VIVO: boolean;

	/**
	 * Running in the huawei's quick game.
	 */
	export const HUAWEI: boolean;

	/**
	 * Running in the cocosplay.
	 */
	export const COCOSPLAY: boolean;

	/**
	 * Running in the qtt's quick game.
	 */
	export const QTT: boolean;

	/**
	 * Running in the linksure's quick game.
	 */
	export const LINKSURE: boolean;

	/**
	 * Running in the editor.
	 */
	export const EDITOR: boolean;

	/**
	 * Preview in browser or simulator.
	 */
	export const PREVIEW: boolean;

	/**
	 * Running in published project.
	 */
	export const BUILD: boolean;

	/**
	 * Running in the engine's unit test.
	 */
	export const TEST: boolean;

	/**
	 * Running debug mode.
	 */
	export const DEBUG: boolean;

	/**
	 * Running in the editor or preview.
	 */
	export const DEV: boolean;

	/**
	 * Running in mini game.
	 */
	export const MINIGAME: boolean;

	/**
	 * Running in runtime based environment.
	 */
	export const RUNTIME_BASED: boolean;

	/**
	 * Support JIT.
	 */
	export const SUPPORT_JIT: boolean;

	/**
	 * Running in environment where using JSB as the JavaScript interface binding scheme.
	 */
	export const JSB: boolean;

	/**
	 * The network access mode.
	 * - 0 Client
	 * - 1 ListenServer
	 * - 2 HostServer
	 */
	export const NET_MODE: number;

	/**
	 * Run in editor but not in editor preview.
	 */
	export const EDITOR_NOT_IN_PREVIEW: boolean;

}
"
`;

exports[`generateInternalConstants 1`] = `
"declare module 'internal:constants'{
	/**
	 * Running in Web platform
	 */
	export const HTML5: boolean;

	/**
	 * Running in native platform (mobile app, desktop app, or simulator).
	 */
	export const NATIVE: boolean;

	/**
	 * Running in ANDROID platform
	 */
	export const ANDROID: boolean;

	/**
	 * Running in IOS platform
	 */
	export const IOS: boolean;

	/**
	 * Running in MAC platform
	 */
	export const MAC: boolean;

	/**
	 * Running in WINDOWS platform
	 */
	export const WINDOWS: boolean;

	/**
	 * Running in LINUX platform
	 */
	export const LINUX: boolean;

	/**
	 * Running in OHOS platform
	 */
	export const OHOS: boolean;

	/**
	 * Running in OPEN_HARMONY platform
	 */
	export const OPEN_HARMONY: boolean;

	/**
	 * Running in the Wechat's mini game.
	 */
	export const WECHAT: boolean;

	/**
	 * Running in the Wechat's mini program.
	 */
	export const WECHAT_MINI_PROGRAM: boolean;

	/**
	 * Running in the baidu's mini game.
	 */
	export const BAIDU: boolean;

	/**
	 * Running in the xiaomi's quick game.
	 */
	export const XIAOMI: boolean;

	/**
	 * Running in the alipay's mini game.
	 */
	export const ALIPAY: boolean;

	/**
	 * Running in the taobao creative app.
	 */
	export const TAOBAO: boolean;

	/**
	 * Running in the taobao mini game.
	 */
	export const TAOBAO_MINIGAME: boolean;

	/**
	 * Running in the ByteDance's mini game.
	 */
	export const BYTEDANCE: boolean;

	/**
	 * Running in the oppo's quick game.
	 */
	export const OPPO: boolean;

	/**
	 * Running in the vivo's quick game.
	 */
	export const VIVO: boolean;

	/**
	 * Running in the huawei's quick game.
	 */
	export const HUAWEI: boolean;

	/**
	 * Running in the cocosplay.
	 */
	export const COCOSPLAY: boolean;

	/**
	 * Running in the qtt's quick game.
	 */
	export const QTT: boolean;

	/**
	 * Running in the linksure's quick game.
	 */
	export const LINKSURE: boolean;

	/**
	 * Running in the editor.
	 */
	export const EDITOR: boolean;

	/**
	 * Preview in browser or simulator.
	 */
	export const PREVIEW: boolean;

	/**
	 * Running in published project.
	 */
	export const BUILD: boolean;

	/**
	 * Running in the engine's unit test.
	 */
	export const TEST: boolean;

	/**
	 * Running debug mode.
	 */
	export const DEBUG: boolean;

	/**
	 * Running in the server mode.
	 */
	export const SERVER_MODE: boolean;

	/**
	 * Running in the editor or preview.
	 */
	export const DEV: boolean;

	/**
	 * Running in mini game.
	 */
	export const MINIGAME: boolean;

	/**
	 * Running in runtime based environment.
	 */
	export const RUNTIME_BASED: boolean;

	/**
	 * Support JIT.
	 */
	export const SUPPORT_JIT: boolean;

	/**
	 * Running in environment where using JSB as the JavaScript interface binding scheme.
	 */
	export const JSB: boolean;

	/**
	 * This is an internal constant to determine whether pack physx libs.
	 */
	export const NOT_PACK_PHYSX_LIBS: boolean;

	/**
	 * The network access mode.
	 * - 0 Client
	 * - 1 ListenServer
	 * - 2 HostServer
	 */
	export const NET_MODE: number;

	/**
	 * Running with webgpu rendering backend.
	 */
	export const WEBGPU: boolean;

	/**
	 * Native code (wasm/asmjs) bundle mode, 0: asmjs, 1: wasm, 2: both
	 */
	export const NATIVE_CODE_BUNDLE_MODE: number;

	/**
	 * Run in editor but not in editor preview.
	 */
	export const EDITOR_NOT_IN_PREVIEW: boolean;

	/**
	 * An internal constant to indicate whether we use wasm assets as minigame subpackage.
	 * This is useful when we need to reduce code size.
	 */
	export const WASM_SUBPACKAGE: boolean;

	/**
	 * An internal constant to indicate whether we cull the meshopt wasm module and asm.js module.
	 */
	export const CULL_MESHOPT: boolean;

}
"
`;
